### PR TITLE
refactor(experience): experience api migration

### PR DESCRIPTION
**Pull request by simeng-li**
## Summary

This PR combines all the experience API migration PRs into one. Includes the full migration code changes. 

Migrate our existing experience app using the latest experience API from legacy interaction API. 

## Testing

Unit test + Integration test

## Checklist

- [ ] `.changeset`
- [x] unit tests
- [x] integration tests
- [x] necessary TSDoc comments

### CODE REVIEW

File: packages/experience/src/apis/experience.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  type PasswordVerificationPayload,
+  SignInIdentifier,
+  type UpdateProfileApiPayload,
+} from '@logto/schemas';
+
+import api from './api';
+
+const prefix = '/api/experience';
+
+const experienceRoutes = Object.freeze({
```
Comment (xiaoyijun) [inline packages/experience/src/apis/experience.ts:13]: We have an `experience.routes` constant in the `schema` package, maybe use `experienceApiRoutes` or something.

File: packages/experience/src/apis/experience.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    json: {
+      interactionEvent,
+    },
+  });
+
+const identifyUser = async (payload: IdentificationApiPayload = {}) =>
+  api.post(experienceRoutes.identification, { json: payload });
+
+const submitInteraction = async () =>
+  api.post(`${experienceRoutes.prefix}/submit`).json<SubmitInteractionResponse>();
```
Comment (xiaoyijun) [inline packages/experience/src/apis/experience.ts:40]:
```suggestion
  api.post(`${experienceRoutes.submit}`).json<SubmitInteractionResponse>();
```
add `submit` to the constant?

File: packages/experience/src/hooks/use-password-policy-checker.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import { useCallback } from 'react';
+
+import usePasswordErrorMessage from './use-password-error-message';
+import { usePasswordPolicy } from './use-sie';
+
+type PasswordPolicyCheckProps = {
```
Comment (xiaoyijun) [inline packages/experience/src/hooks/use-password-policy-checker.ts:6]: nit: I notice that in our code base, we always use `...Props` for react components, and use `...Options` for hooks. Should we stay consistent with the previous code?

File: packages/experience/src/hooks/use-password-rejection-handler.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
+import { type RequestErrorBody } from '@logto/schemas';
+import { useCallback, useMemo } from 'react';
+
+import type { ErrorHandlers } from './use-error-handler';
+import usePasswordErrorMessage from './use-password-error-message';
+
+type ErrorHandlerProps = {
```
Comment (xiaoyijun) [inline packages/experience/src/hooks/use-password-rejection-handler.ts:7]: ditto

File: packages/experience/src/apis/experience.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+const updateProfile = async (payload: UpdateProfileApiPayload) => {
+  await api.post(experienceRoutes.profile, { json: payload });
+};
+
+export const signInWithPasswordIdentifier = async (payload: PasswordVerificationPayload) => {
+  await initInteraction(InteractionEvent.SignIn);
```
Comment (xiaoyijun) [inline packages/experience/src/apis/experience.ts:47]: Will this step returns an error？

Comment (xiaoyijun) [reply packages/experience/src/apis/experience.ts:47]: ditto for the following code:
```ts
await identifyUser({ verificationId });
```

Comment (simeng-li) [reply packages/experience/src/apis/experience.ts:47]: Yes, and it will be caught and returned to the upper scope.

**Review by xiaoyijun — APPROVED** [review]

**Review by xiaoyijun — CHANGES_REQUESTED** [review]
Block manually

File: .changeset/fresh-shrimps-rhyme.md [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+---
+
+migrate experience app using Experience API.
+
+Migrate experience app API requests from legacy [Interaction API](https://openapi.logto.io/group/endpoint-interaction) to the new [Experience API](https://openapi.logto.io/group/endpoint-experience), except the following endpoints:
+
+- `GET /api/interaction/consent`
+- `POST /api/interaction/consent`
+
+Those endpoints are used in the third-party application's consent page only. Remain unchanged.
```
Comment (gao-sun) [inline .changeset/fresh-shrimps-rhyme.md:12]: can we rename to `api/experience/consent`?

File: packages/experience/src/apis/const.ts [hunk 7]
```diff
@@ -0,0 +1 @@
+export const kyPrefixUrl = '/';
```
Comment (gao-sun) [inline packages/experience/src/apis/const.ts:1]: not in use?

File: packages/experience/src/apis/experience/index.ts [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+  submitInteraction,
+  identifyUser,
+  identifyAndSubmitInteraction,
+} from './interaction';
+
+export * from './mfa';
+export * from './social';
+
+export const registerWithVerifiedIdentifier = async (verificationId: string) => {
+  await updateInteractionEvent(InteractionEvent.Register);
```
Comment (gao-sun) [inline packages/experience/src/apis/experience/index.ts:33]: can we make an experience api SDK and use it for all API calls later?

File: packages/experience/src/apis/experience/interaction.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    },
+  });
+
+export const identifyUser = async (payload: IdentificationApiPayload = {}) =>
+  api.post(experienceApiRoutes.identification, { json: payload });
+
+export const submitInteraction = async () =>
+  api.post(`${experienceApiRoutes.submit}`).json<SubmitInteractionResponse>();
+
+export const _updateProfile = async (payload: UpdateProfileApiPayload) =>
```
Comment (gao-sun) [inline packages/experience/src/apis/experience/interaction.ts:28]: why `_`?

File: packages/experience/src/apis/experience/mfa.ts [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+} from '@logto/schemas';
+
+import api from '../api';
+
+import { experienceApiRoutes } from './const';
+import { submitInteraction } from './interaction';
+
+/**
+ * Mfa APIs
+ */
```
Comment (gao-sun) [inline packages/experience/src/apis/experience/mfa.ts:16]:
```suggestion
```

File: packages/experience/src/apis/experience/social.ts [hunk 11]
```diff
@@ -0,0 +1,96 @@
+// Social and SSO APIs
```
Comment (gao-sun) [inline packages/experience/src/apis/experience/social.ts:1]: this looks like a stop sign comment if you can rename the file to `social-and-sso`

File: packages/experience/src/apis/utils.ts [hunk 12]
```diff
@@ -1,26 +1,51 @@
...
+      return sendVerificationCode(InteractionEvent.SignIn, identifier);
+    }
+    case UserFlow.Register: {
+      await initInteraction(InteractionEvent.Register);
+      return sendVerificationCode(InteractionEvent.Register, identifier);
+    }
+    case UserFlow.ForgotPassword: {
+      await initInteraction(InteractionEvent.ForgotPassword);
+      return sendVerificationCode(InteractionEvent.ForgotPassword, identifier);
+    }
```
Comment (gao-sun) [inline packages/experience/src/apis/utils.ts:25]: can we merge them into one? e.g. mapping `UserFlow` to `InteractionEvent`

File: packages/experience/src/apis/utils.ts [hunk 13]
```diff
@@ -1,26 +1,51 @@
...
+  switch (type) {
+    case UserFlow.SignIn: {
+      return sendVerificationCode(InteractionEvent.SignIn, identifier);
+    }
+    case UserFlow.Register: {
+      return sendVerificationCode(InteractionEvent.Register, identifier);
+    }
+    case UserFlow.ForgotPassword: {
+      return sendVerificationCode(InteractionEvent.ForgotPassword, identifier);
+    }
```
Comment (gao-sun) [inline packages/experience/src/apis/utils.ts:45]: ditto

File: packages/experience/src/containers/VerificationCode/use-continue-flow-code-verification.ts [hunk 14]
```diff
@@ -1,74 +1,88 @@
...
-        await showLinkSocialConfirmModal(method, target, linkSocial);
-
-        return;
-      }
-
-      await showIdentifierErrorAlert(IdentifierErrorType.IdentifierAlreadyExists, method, target);
-    },
-    [searchParameters, showIdentifierErrorAlert, showLinkSocialConfirmModal]
-  );
+  const identifierExistErrorHandler = useCallback(async () => {
```
Comment (gao-sun) [inline packages/experience/src/containers/VerificationCode/use-continue-flow-code-verification.ts:45]:
```suggestion
  const identifierExistsErrorHandler = useCallback(async () => {
```

File: packages/experience/src/hooks/use-start-backup-code-binding.ts [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+import { type BackupCodeBindingState } from '@/types/guard';
+
+import useApi from './use-api';
+import useErrorHandler from './use-error-handler';
+
+type Options = {
+  replace?: boolean;
+};
+
+const useBackupCodeBinding = ({ replace }: Options = {}) => {
```
Comment (gao-sun) [inline packages/experience/src/hooks/use-start-backup-code-binding.ts:17]: function name and file name mismatch

File: packages/experience/src/hooks/use-start-backup-code-binding.ts [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+import UserInteractionContext from '@/Providers/UserInteractionContextProvider/UserInteractionContext';
+import { createBackupCode } from '@/apis/experience';
+import { UserMfaFlow } from '@/types';
+import { type BackupCodeBindingState } from '@/types/guard';
+
+import useApi from './use-api';
+import useErrorHandler from './use-error-handler';
+
+type Options = {
+  replace?: boolean;
```
Comment (gao-sun) [inline packages/experience/src/hooks/use-start-backup-code-binding.ts:14]: this is confusing, my first impression is "replace backup codes". can you add some comments?

File: packages/experience/src/hooks/use-webauthn-operation.ts [hunk 17]
```diff
@@ -63,19 +63,26 @@ const useWebAuthnOperation = () => {
...
+          ? {
+              flow: UserMfaFlow.MfaVerification,
+              payload: { ...response, type: MfaFactor.WebAuthn },
+              verificationId,
+            }
+          : {
+              flow: UserMfaFlow.MfaBinding,
+              payload: { ...response, type: MfaFactor.WebAuthn },
+              verificationId,
+            }
```
Comment (gao-sun) [inline packages/experience/src/hooks/use-webauthn-operation.ts:84]:
```ts
{
  flow: isAuthenticationResponseJSON(response)
    ? UserMfaFlow.MfaVerification : UserMfaFlow.MfaBinding,
  payload: { ...response, type: MfaFactor.WebAuthn },
  verificationId,
}
```

**Review by gao-sun — CHANGES_REQUESTED** [review]

File: .changeset/fresh-shrimps-rhyme.md (hunk 6 above)
Comment (simeng-li) [reply .changeset/fresh-shrimps-rhyme.md:12]: Have an issue tracking this, will migrate these two APIs under the experience routes.

File: packages/experience/src/apis/experience/index.ts (hunk 8 above)
Comment (simeng-li) [reply packages/experience/src/apis/experience/index.ts:33]: Yes,  create an issue tracking this. Consider gradually migrating and maybe with new elements based components.

File: packages/experience/src/apis/experience/interaction.ts (hunk 9 above)
Comment (simeng-li) [reply packages/experience/src/apis/experience/interaction.ts:28]: This is for private internal API reference use only. Have another `updateProfile` API exported to the web components. 

image

Comment (simeng-li) [reply packages/experience/src/apis/experience/interaction.ts:28]: Updated change the API name:

`_updateProfile` -> `updateProfile`
`updateProfile` -> `fulfillProfile`

File: packages/experience/src/apis/experience/social.ts (hunk 11 above)
Comment (simeng-li) [reply packages/experience/src/apis/experience/social.ts:1]: Yes, those are actually legacy comments, originally they all come from a single file. Will clean them up.

File: packages/experience/src/apis/utils.ts (hunk 12 above)
Comment (simeng-li) [reply packages/experience/src/apis/utils.ts:25]: That is a legacy issue, we had a `UserFlow.Continue`.  `UserFlow` can not be strictly mapped to `InteractionEvent`.  

I'll create an issue for this. Needs some refactor work.

Comment (simeng-li) [reply packages/experience/src/apis/utils.ts:25]: For now, create a map for it

File: packages/experience/src/hooks/use-webauthn-operation.ts (hunk 17 above)
Comment (simeng-li) [reply packages/experience/src/hooks/use-webauthn-operation.ts:84]: Had a similar concern. Looks like the input type is a union, this shorthand code can not infer types properly.

File: packages/experience/src/pages/SocialSignInWebCallback/use-social-sign-in-listener.ts [hunk 18]
```diff
@@ -59,17 +81,30 @@ const useSocialSignInListener = (connectorId: string) => {
...
       connectorId,
       navigate,
       registerWithSocial,
+      setToast,
       socialSignInSettings.automaticAccountLinking,
+      t,
     ]
   );
 
+  const globalErrorHandler = useMemo<ErrorHandlers>(
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/experience/src/pages/SocialSignInWebCallback/use-social-sign-in-listener.ts:97]: The return value is an object.

File: packages/experience/src/apis/utils.ts (hunk 13 above)
Comment (simeng-li) [reply packages/experience/src/apis/utils.ts:45]: updated

File: packages/experience/src/pages/SocialLinkAccount/index.tsx [hunk 19]
```diff
@@ -45,11 +49,19 @@ const SocialLinkAccount = () => {
     return <ErrorPage rawMessage="Connector not found" />;
   }
 
+  if (!verificationId) {
+    return <ErrorPage title="error.invalid_session" rawMessage="Verification id not found" />;
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/experience/src/pages/SocialLinkAccount/index.tsx]: I think all these error messages are edge cases, so no i18n is implemented for those errors.

File: packages/experience/src/pages/VerificationCode/index.tsx [hunk 20]
```diff
@@ -33,19 +44,24 @@ const VerificationCode = () => {
...
   // SignIn Method not enabled
+  const methodSettings = signInMethods.find((method) => method.identifier === type);
   if (!methodSettings && flow !== UserFlow.ForgotPassword) {
     return <ErrorPage />;
   }
 
+  // VerificationId not found
+  const verificationId = verificationIdsMap[codeVerificationTypeMap[type]];
+  if (!verificationId) {
+    return <ErrorPage title="error.invalid_session" rawMessage="Verification id not found" />;
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/experience/src/pages/VerificationCode/index.tsx]: ? Not following

File: packages/experience/src/pages/SocialLinkAccount/index.tsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (gao-sun) [reply packages/experience/src/pages/SocialLinkAccount/index.tsx]: Edge cases are still cases. Good for now.

File: packages/experience/src/pages/VerificationCode/index.tsx (hunk 20 above)
[the remark this answers is not in the document]
Comment (gao-sun) [reply packages/experience/src/pages/VerificationCode/index.tsx]: no i18n. good for now

File: packages/experience/src/apis/experience/index.ts (hunk 8 above)
Comment (gao-sun) [reply packages/experience/src/apis/experience/index.ts:33]: can you paste the issue number here?

File: .changeset/fresh-shrimps-rhyme.md (hunk 6 above)
Comment (gao-sun) [reply .changeset/fresh-shrimps-rhyme.md:12]: can you paste the issue number here?

File: packages/experience/src/hooks/use-start-backup-code-binding.ts [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+import UserInteractionContext from '@/Providers/UserInteractionContextProvider/UserInteractionContext';
+import { createBackupCode } from '@/apis/experience';
+import { UserMfaFlow } from '@/types';
+import { type BackupCodeBindingState } from '@/types/guard';
+
+import useApi from './use-api';
+import useErrorHandler from './use-error-handler';
+
+type Options = {
+  /** Whether to replace the current page in the history stack. */
```
Comment (gao-sun) [inline packages/experience/src/hooks/use-start-backup-code-binding.ts:14]: still confusing.

this hook is returning a function, and this option is affecting the behavior in that function - should specify this or remove this param

File: packages/experience/src/hooks/use-webauthn-operation.ts (hunk 17 above)
Comment (gao-sun) [reply packages/experience/src/hooks/use-webauthn-operation.ts:84]: ok then

File: packages/experience/src/pages/SocialSignInWebCallback/use-social-sign-in-listener.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (gao-sun) [reply packages/experience/src/pages/SocialSignInWebCallback/use-social-sign-in-listener.ts:97]: the object only has a function, and the name "handler" suggests it's a function

File: packages/experience/src/apis/experience/index.ts (hunk 8 above)
Comment (simeng-li) [reply packages/experience/src/apis/experience/index.ts:33]: LOG-9941

File: packages/experience/src/pages/SocialSignInWebCallback/use-social-sign-in-listener.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/experience/src/pages/SocialSignInWebCallback/use-social-sign-in-listener.ts:97]: make sense, updated.

File: packages/experience/src/hooks/use-start-backup-code-binding.ts (hunk 21 above)
Comment (simeng-li) [reply packages/experience/src/hooks/use-start-backup-code-binding.ts:14]: Update all legacy hooks, remove this param from hooks. Should pass it in the callback method directly.

**Review by gao-sun — APPROVED** [review]

**Review by xiaoyijun — APPROVED** [review]

**Pull request merged**